controllers/harbor/provision.go

```diff
@@ -0,0 +1,307 @@
+				NodeSelector:     nil,
+				ImagePullSecrets: harbor.getImagePullSecrets(),
+			},
+			RedisSecret: harbor.getJobServiceCacheSecret(),
+			WorkerCount: harbor.HarborCluster.Spec.JobService.WorkerCount,
+		}
+	}
+	return nil
+}
+
+func (harbor *HarborReconciler) newChartMuseumComponentIfNecessary() *v1alpha1.ChartMuseumComponent {
+	if harbor.HarborCluster.Spec.ChartMuseum != nil {
+		return &v1alpha1.ChartMuseumComponent{
+			HarborDeployment: v1alpha1.HarborDeployment{
+				Replicas:         IntToInt32Ptr(harbor.HarborCluster.Spec.Replicas),
+				Image:            harbor.ImageGetter.ChartMuseumImage(),
+				NodeSelector:     nil,
+				ImagePullSecrets: harbor.getImagePullSecrets(),
+			},
+			StorageSecret: harbor.getStorageSecret(),
+			CacheSecret:   harbor.getChartMuseumCacheSecret(),
+		}
+	}
+	return nil
+}
+
+func (harbor *HarborReconciler) newClairComponentIfNecessary() *v1alpha1.ClairComponent {
+	if harbor.HarborCluster.Spec.Clair != nil {
+		return &v1alpha1.ClairComponent{
+			HarborDeployment: v1alpha1.HarborDeployment{
+				Replicas:         IntToInt32Ptr(harbor.HarborCluster.Spec.Replicas),
+				Image:            harbor.ImageGetter.ClairImage(),
+				NodeSelector:     nil,
+				ImagePullSecrets: harbor.getImagePullSecrets(),
+			},
+			DatabaseSecret:       harbor.getClairDatabaseSecret(),
+			VulnerabilitySources: harbor.HarborCluster.Spec.Clair.VulnerabilitySources,
+			Adapter: v1alpha1.ClairAdapterComponent{
+				Image:       harbor.ImageGetter.ClairAdapterImage(),
+				RedisSecret: harbor.getClairCacheSecret(),
+			},
+		}
+	}
+	return nil
+}
+
+// newNotaryComponentIfNecessary will return a NotaryComponent in harbor CRD.
+func (harbor *HarborReconciler) newNotaryComponentIfNecessary() *v1alpha1.NotaryComponent {
+	if harbor.HarborCluster.Spec.Notary != nil {
+		return &v1alpha1.NotaryComponent{
+			PublicURL: harbor.HarborCluster.Spec.Notary.PublicURL,
+			DBMigrator: v1alpha1.NotaryDBMigrator{
+				Image: harbor.ImageGetter.NotaryDBMigratorImage(),
+			},
+			Signer: v1alpha1.NotarySignerComponent{
+				HarborDeployment: v1alpha1.HarborDeployment{
+					Replicas:         IntToInt32Ptr(harbor.HarborCluster.Spec.Replicas),
+					Image:            harbor.ImageGetter.NotarySingerImage(),
+					NodeSelector:     nil,
+					ImagePullSecrets: harbor.getImagePullSecrets(),
+				},
+				DatabaseSecret: harbor.getNotarySignerDatabaseSecret(),
+			},
+			Server: v1alpha1.NotaryServerComponent{
+				HarborDeployment: v1alpha1.HarborDeployment{
+					Replicas:         IntToInt32Ptr(harbor.HarborCluster.Spec.Replicas),
+					Image:            harbor.ImageGetter.NotaryServerImage(),
+					NodeSelector:     nil,
+					ImagePullSecrets: harbor.getImagePullSecrets(),
+				},
+				DatabaseSecret: harbor.getNotaryServerDatabaseSecret(),
+			},
+		}
+	}
+	return nil
+}
+
+// getRegistryCacheSecret will get a name of k8s secret which stores cache info
+func (harbor *HarborReconciler) getRegistryCacheSecret() string {
```

Comment:
IMO, defining so many helper functions does not bring to many simplifications. Maybe `harbor.getStorageProperty` is enough simple to call.

Reply:
Code shown here seems a little redundant.